config-model/src/main/java/com/yahoo/searchdefinition/processing/OnnxModelTypeResolver.java

```diff
@@ -0,0 +1,154 @@
+                    modelConfig.addInputNameMapping(onnxInputName, vespaInputName, false);
+                    modelConfig.addInputType(onnxInputName, valueInfo.getType());
+                }
+
+                // Model outputs
+                for (Onnx.ValueInfoProto valueInfo : model.getGraph().getOutputList()) {
+                    String onnxOutputName = valueInfo.getName();
+                    String vespaOutputName = OnnxModelTransformer.asValidIdentifier(onnxOutputName);
+                    modelConfig.addOutputNameMapping(onnxOutputName, vespaOutputName, false);
+                    modelConfig.addOutputType(onnxOutputName, valueInfo.getType());
+                }
+
+                // Set the first output as default
+                if ( ! model.getGraph().getOutputList().isEmpty()) {
+                    modelConfig.setDefaultOutput(model.getGraph().getOutput(0).getName());
+                }
+
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Unable to parse ONNX model", e);
+            }
+        }
+    }
+
+    static boolean modelFileExists(String path, ApplicationPackage app) {
+        Path pathInApplicationPackage = Path.fromString(path);
+        if (getFile(pathInApplicationPackage, app).exists()) {
+            return true;
+        }
+        if (getFileReference(pathInApplicationPackage, app).isPresent()) {
+            return true;
+        }
+        return false;
+    }
+
+    private InputStream openModelFile(Path path) throws FileNotFoundException {
+        ApplicationFile file;
+        Optional<FileReference> reference;
+        Path modelsPath = ApplicationPackage.MODELS_DIR.append(path);
+
+        if ((file = getFile(path)).exists()) {
+            return file.createInputStream();
+        }
+        if ((file = getFile(modelsPath)).exists()) {
+            return file.createInputStream();
+        }
+        if ((reference = getFileReference(path)).isPresent()) {
+            return openFromFileRepository(path, reference.get());
+        }
+        if ((reference = getFileReference(modelsPath)).isPresent()) {
+            return openFromFileRepository(modelsPath, reference.get());
+        }
+
+        throw new IllegalArgumentException("Unable to find ONNX model file \"" + path + "\" " +
+            "in application package or file repository.");
+    }
+
+    private ApplicationFile getFile(Path path) {
+        return getFile(path, search.applicationPackage());
+    }
+
+    private static ApplicationFile getFile(Path path, ApplicationPackage app) {
+        return app.getFile(path);
+    }
+
+    private static InputStream openFromFileRepository(Path path, FileReference reference) throws FileNotFoundException {
+        return new FileInputStream(new File(getFileRepositoryPath(path, reference.value())));
+    }
+
+    public static String getFileRepositoryPath(Path path, String fileReference) {
```

Comment:
This seems like a hack to me but I guess it will work. I think we should be able to ask the application package to give us the file regardless of how it is stored in the implementation class. 

There's probably more than enough in the PR already, but we should follow it up somehow. @hmusum any thoughts on this?

Reply:
I think we can just drop the `getFileReference` (and subsequent `openFromFileRepository`) calls. If `getFile` fails `getFileReference` will also fail. If there exists is a file reference for a file in the application package it will be added to the application package directory immediately afterwards and it will be found when doing `getFile`.

Reply:
So, the reason this is here is the `getFile` fails to find the file when the application package is backed by `ZKApplicationPackage` because the actual file hasn't been added there. See `ZooKeeperClient.java`. During the prepare step, as far as I understand, at first the application package is a `FilesApplicationPackage`. Here the file exists. However, during config deriving, the ZK application package will be created, and the model file will not be put there, in the `.preprocessed` directory. However, we still require the actual file. That's why we look in the file repository, and at this stage the file reference exists and the file has been added there.

Now, I'm not sure what happens with this scenario when we have multiple config servers. I guess we'll just have to try?